// pca9685/dist/src/pca9685.js
/*
 * src/pca9685.ts
 * https://github.com/101100/pca9685
 *
 * Library for PCA9685 I2C 16-channel PWM/servo driver.
 *
 * Copyright (c) 2015 Jason Heard
 * Licensed under the MIT license.
 */
"use strict";
var debugFactory = require("debug");
var constants = {
    modeRegister1: 0x00,
    modeRegister1Default: 0x01,
    sleepBit: 0x10,
    restartBit: 0x80,
    modeRegister2: 0x01,
    modeRegister2Default: 0x04,
    channel0OnStepLowByte: 0x06,
    channel0OnStepHighByte: 0x07,
    channel0OffStepLowByte: 0x08,
    channel0OffStepHighByte: 0x09,
    registersPerChannel: 4,
    allChannelsOnStepLowByte: 0xFA,
    allChannelsOnStepHighByte: 0xFB,
    allChannelsOffStepLowByte: 0xFC,
    allChannelsOffStepHighByte: 0xFD,
    preScale: 0xFE,
    stepsPerCycle: 4096,
    defaultAddress: 0x40,
    defaultFrequency: 50,
    baseClockHertz: 25000000
};
function createSetFrequencyStep2(sendFunc, debug, prescale, cb) {
    cb = typeof cb === "function" ? cb : function () { return; };
    return function setFrequencyStep2(err, byte) {
        if (err) {
            debug("Error reading mode (to set frequency)", err);
            cb(err);
        }
        var oldmode = byte;
        var newmode = (oldmode & ~constants.restartBit) | constants.sleepBit;
        debug("Setting prescale to: %d", prescale);
        sendFunc(constants.modeRegister1, newmode);
        sendFunc(constants.preScale, Math.floor(prescale));
        sendFunc(constants.modeRegister1, oldmode);
        // documentation says that 500 microseconds are required
        // before restart is sent, so a timeout of 10 milliseconds
        // should be plenty
        setTimeout(function () {
            debug("Restarting controller");
            sendFunc(constants.modeRegister1, oldmode | constants.restartBit);
            cb();
        }, 10);
    };
}
var Pca9685Driver = (function () {
    /**
     * Constructs a
     *
     * @param {Pca9685Options}
     * @param {any)        => any}
     */
    function Pca9685Driver(options, cb) {
        if (options.debug) {
            debugFactory.enable("pca9685");
        }
        this.i2c = options.i2c;
        this.address = options.address || constants.defaultAddress;
        this.debug = debugFactory("pca9685");
        this.frequency = options.frequency || constants.defaultFrequency;
        var cycleLengthMicroSeconds = 1000000 / this.frequency;
        this.stepLengthMicroSeconds = cycleLengthMicroSeconds / constants.stepsPerCycle;
        this.send = this.send.bind(this);
        this.debug("Reseting PCA9685");
        this.send(constants.modeRegister1, constants.modeRegister1Default);
        this.send(constants.modeRegister2, constants.modeRegister2Default);
        this.allChannelsOff();
        this.setFrequency(this.frequency, cb);
    }
    /**
     * Sets the on and off steps for the given channel.
     *
     * @param {number} channel
     *     Output hannel to configure.
     * @param {number} onStep
     *     The step number when the channel should turn on.
     * @param {number} offStep
     *     The step number when the channel should turn off.
     */
    Pca9685Driver.prototype.setPulseRange = function (channel, onStep, offStep) {
        this.debug("Setting PWM channel, channel: %d, onStep: %d, offStep: %d", channel, onStep, offStep);
        this.send(constants.channel0OnStepLowByte + constants.registersPerChannel * channel, onStep & 0xFF);
        this.send(constants.channel0OnStepHighByte + constants.registersPerChannel * channel, (onStep >> 8) & 0x0F);
        this.send(constants.channel0OffStepLowByte + constants.registersPerChannel * channel, offStep & 0xFF);
        this.send(constants.channel0OffStepHighByte + constants.registersPerChannel * channel, (offStep >> 8) & 0x0F);
    };
    /**
     * Sets the pulse length for the given channel.
     *
     * @param {number} channel
     *     Output hannel to configure.
     * @param {number} pulseLengthMicroSeconds
     *     The length of the pulse for the given channel in microseconds.
     * @param {number} onStep
     *     Optional The step number when the channel should turn on (defaults
     *     to 0).
     */
    Pca9685Driver.prototype.setPulseLength = function (channel, pulseLengthMicroSeconds, onStep) {
        if (onStep === void 0) { onStep = 0; }
        this.debug("Setting PWM channel, channel: %d, pulseLength: %d, onStep: %d", channel, pulseLengthMicroSeconds, onStep);
        var offStep = (onStep + Math.round(pulseLengthMicroSeconds / this.stepLengthMicroSeconds) - 1) % constants.stepsPerCycle;
        this.setPulseRange(channel, onStep, offStep);
    };
    /**
     * Sets the duty cycle for the given channel.
     *
     * @param {number} channel
     *     Output hannel to configure.
     * @param {number} dutyCycleDecimalPercentage
     *     The duty cycle for the given channel as a decimal percentage.
     * @param {number} onStep
     *     Optional The step number when the channel should turn on (defaults
     *     to 0).
     */
    Pca9685Driver.prototype.setDutyCycle = function (channel, dutyCycleDecimalPercentage, onStep) {
        if (onStep === void 0) { onStep = 0; }
        this.debug("Setting PWM channel, channel: %d, dutyCycle: %f, onStep: %d", channel, dutyCycleDecimalPercentage, onStep);
        var offStep = (onStep + Math.round(dutyCycleDecimalPercentage * constants.stepsPerCycle) - 1) % constants.stepsPerCycle;
        this.setPulseRange(channel, onStep, offStep);
    };
    /**
     * Turns all channels off.
     */
    Pca9685Driver.prototype.allChannelsOff = function () {
        // Setting the high byte to 1 will turn off the channel
        this.send(constants.allChannelsOffStepHighByte, 0x01);
    };
    Pca9685Driver.prototype.setFrequency = function (freq, cb) {
        // 25MHz base clock, 12 bit (4096 steps per cycle)
        var prescale = Math.round(constants.baseClockHertz / (constants.stepsPerCycle * freq)) - 1;
        this.debug("Setting PWM frequency to %d Hz", freq);
        this.debug("Pre-scale value: %d", prescale);
        this.i2c.readByte(this.address, constants.modeRegister1, createSetFrequencyStep2(this.send, this.debug, prescale, cb));
    };
    Pca9685Driver.prototype.send = function (cmd, byte) {
        this.i2c.writeByte(this.address, cmd, byte, function (err) {
            if (err) {
                console.log("Error writing to PCA8685 via I2C", err);
            }
        });
    };
    return Pca9685Driver;
}());
exports.Pca9685Driver = Pca9685Driver;
